### PR TITLE
Improve Amphp performance

### PR DESCRIPTION
**Pull request by dbalabka**
Amphp should implement queuing approach similar to React. 
WIP. It didn't yet perform manual testing. So not sure that this works.

### CODE REVIEW

**Comment by Insolita** [comment]
Hi, thanks a lot!

**Comment by Insolita** [comment]
Merged to the "[next](https://github.com/Insolita/php-async-benchmarks/commit/2a4906ecde375ec6dfae537a1ea1babca46e09ab)" branch

**Comment by Insolita** [comment]
Huh, make it workable https://github.com/Insolita/php-async-benchmarks/blob/next/src/clients/Amphp.php

But it still slow.   ¯\_(ツ)_/¯
Probably I haven't some of the necessary libs?

**Comment by kelunik** [comment]
The queue is built-in depending on the pool you choose. I'll send a PR. :-)

**Comment by dbalabka** [comment]
@Insolita I have adjusted code and fix the bugs. 

@kelunik I tried to get how to use amp/parallel Pools but it is not clear from the documentation. In my case, I have implemented a simple queuing based on SplQueue and pool of workers (promises).

**Comment by kelunik** [comment]
@dbalabka Parallel isn't needed, but `amphp/sync` can be helpful, see #2. It's quite a bit faster than Guzzle for me now.

**Comment by dbalabka** [comment]
@kelunik I just reviewed the code. In my opinion, semaphores do not look so straight forward. Could you please compare my implementation (I created another PR due to this one closed):
https://github.com/Insolita/php-async-benchmarks/pull/3/files 
with yours and provide your feedback. It would be very valuable.

File: src/clients/Amphp.php [hunk 1]
```diff
@@ -57,49 +61,48 @@ private function initUrls()
...
+                $promise = $this->processRequest($this->queue->pop());
+
+                $promise->onResolve(function () use (&$pool) {
+                    unset($pool[array_search($this, $pool, true)]);
+                });
+                $pool[] = $promise;
+                continue;
             }
+            // wait when some task will be accomplished
+            [$errors, $values] = yield Promise\some($pool);
```
Comment (kelunik) [inline src/clients/Amphp.php:88]: Using `some` here will register a callback to each promise every iteration, which will negatively affect performance and memory usage.

Comment (dbalabka) [reply src/clients/Amphp.php:88]: It is fixed. Please see new PR:
https://github.com/Insolita/php-async-benchmarks/pull/3/files

Comment (dbalabka) [reply src/clients/Amphp.php:88]: I use `Promise\first()`, so a similar problem will be there

**Comment by Insolita** [comment]
Oh, guys, thanks a lot!  
I'm confused now...  Both versions work good, but it so different

#2 
kelunik
 #3 
dbalabka

**Comment by kelunik** [comment]
@Insolita Seems like the other clients have similar timing differences in the README?

**Comment by kelunik** [comment]
```
php bin/bench -t=amp --batch=200 --conk=50 --iters=2
0 - 31.3855509758
1 - 10.08562707901
```

I just got this locally, it's likely due to DNS being cached on the second run.

**Comment by kelunik** [comment]
Most time is spend inside DNS / connection logic:

```
cat temp/amp/requests.har| jq '.log.entries[].timings.connect' | datamash mean 1
827,52331606218
```

You can get that `requests.har` by adding `->intercept(new LogHttpArchive($this->tempDir . '/requests.har'))` on `HttpClientBuilder`.

**Pull request closed without merging**